### PR TITLE
modify outputs to use comm

### LINKED ISSUE #2953: outputs widget contextmanager fails on metakernel
This is an issue found in  Calysto/metakernel#217

The context manager in the outputs widget requires a call to get_ipython and therefore fails on non-ipython kernels.  The other mecchanisms for access outputs widget appear to work,

### PR DESCRIPTION
**Pull request by joequant**
This modifies the outputs widget to use comm rather than calling
ipython directly.  This fixes #2953 and Calysto/metakernel#217

### CODE REVIEW

**Comment by jasongrout** [comment]
Thanks!

Adding to the queue for review...

File: ipywidgets/widgets/widget_output.py [hunk 1]
```diff
@@ -108,23 +108,39 @@ def __enter__(self):
...
-        ip = get_ipython()
+        kernel = None
         if etype is not None:
+            ip = get_ipython()
             if ip:
+                kernel = ip
                 ip.showtraceback((etype, evalue, tb), tb_offset=0)
+            elif self.comm is not None and self.comm.kernel is not None:
+                kernel = self.comm.kernel
+                kernel.send_response(kernel.iopub_socket,
```
Comment (martinRenou) [inline ipywidgets/widgets/widget_output.py:130]: Not sure using those kernel APIs is a good idea.

I wonder if simply re-raising the error would work? And let the kernel catch it and publish error traceback. If this doesn't work, I don't know a better solution than the current PR.

Comment (jasongrout) [reply ipywidgets/widgets/widget_output.py:130]: The point behind this code is to publish the error message to the output widget (i.e., before `self.msg_id` is reset below). If we just raised the error, we wouldn't be trapping it for the output widget.

It's debatable whether trapping and showing the error in the output widget is better than raising it to the enclosing scope, but that would also be a change in behavior, and I think this PR is trying to be backwards compatible with the behavior.

**Comment by jasongrout** [comment]
I resolved the conflict with master (this conflicted with https://github.com/jupyter-widgets/ipywidgets/pull/3199).

@martinRenou or @ibdafna (or anyone else) - can you review?

**Comment by ibdafna** [comment]
Will review in the morning!

**Review by ibdafna — APPROVED** [review]
LGTM

**Comment by jasongrout** [comment]
Itay, feel free to merge if you are comfortable with the code.

**Comment by jtpio** [comment]
Looks like the test failure was not relevant, and should be fixed in #3227

**Pull request merged**